src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/StringCache.cs

```diff
@@ -0,0 +1,158 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+#nullable enable
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis.Razor.Serialization.Internal
+{
+    internal class StringCache
+    {
+        private readonly ExfiltratingEqualityComparer _comparer;
+        private readonly HashSet<Entry> _hashSet;
+        private readonly object _lock = new object();
+        private int _capacity;
+
+        public StringCache(int capacity = 1024)
+        {
+            _capacity = capacity;
+            _comparer = new ExfiltratingEqualityComparer();
+            _hashSet = new HashSet<Entry>(_comparer);
+        }
+
+        public int ApproximateSize => _hashSet.Count;
+
+        public string GetOrAddValue(string key)
+        {
+            lock (_lock)
+            {
+                if (key is null)
+                {
+                    throw new ArgumentNullException(nameof(key));
+                }
+
+                if (TryGetValue(key, out var result))
+                {
+                    return result!;
+                }
+                else
+                {
+                    var weakRef = new Entry(key);
+                    _hashSet.Add(weakRef);
+
+                    // Whenever we expand lets clean up dead references
+                    if (_capacity <= _hashSet.Count)
+                    {
+                        Cleanup();
+                        _capacity *= 2;
+                    }
+
+                    return key;
+                }
+            }
+        }
+
+        private void Cleanup()
+        {
+            _hashSet.RemoveWhere((weakRef) =>
+            {
+                return !weakRef.IsAlive;
+            });
+        }
+
+        private bool TryGetValue(string key, out string? value)
+        {
+            if (_hashSet.Contains(new Entry(key)))
+            {
+                var exfiltrator = (ExfiltratingEqualityComparer)_hashSet.Comparer;
+                var val = exfiltrator.LastEqualValue!;
```

Comment:
lol omg the hack, only reason why this can't be `null` is because of the `_lock` in the parent method

Reply:
Yeah, it's... less than ideal, but the only other alternatives I can find are to either upgrade this project (and everything it depends on) to netstandard2.1, or to implement our own entire HashSet which DOES have a Get method.

The fact that this IS so gross and fragile is why I really limited the surface area here to just the `GetOrAdd` method, and everything else stays private.